Documentation for some RBT transform methods

amcastro-tri opened the pull request:
Documents `RBT::transformPoints` and `RBT::transformPointsJacobian`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4311)
<!-- Reviewable:end -->

amcastro-tri commented:
+@edrumwri for feature please.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXq482u2cFgbRZVEl8K:b-o13tgg)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.h, line 465 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqD6udzlxn-3ZBVPCn:-KXqD6udzlxn-3ZBVPCo:b1eqoe0) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L465)):*
> ```objc
> 
>   /// Given a set of points in the frame referenced by @p
>   /// from_body_or_frame_ind this methods returns the same set of points in
> ```

"in described in"

---

*[drake/multibody/rigid_body_tree.h, line 470 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqDD5K3kZqYrBWRzdP:-KXqDD5K3kZqYrBWRzdQ:bidgejq) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L470)):*
> ```objc
>   /// @param[in] points A matrix in R^{3 x n} containing n 3D column vectors
>   /// for each point to be transformed.
>   /// @param[in] from_body_or_frame_ind Frame identifier on which input points
> ```

Imagine that I'm looking up this function and I see that it takes an int for a body / frame identifier. What are valid values for that int? What are the meanings of the values? (0 refers to frame what? 1 refers to frame what?). Answering this might also answer my question: why does this method exist here? Put another way: points can be transformed between frames without needing to know anything about rigid bodies. So why is this function in RigidBodyTree? 

---

*[drake/multibody/rigid_body_tree.h, line 474 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqD2egJCIsi4He9ovu:-KXqD2egJCIsi4He9ovv:bc3rtmy) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L474)):*
> ```objc
>   /// @param[in] to_body_or_frame_ind Frame identifier on which output points
>   /// are described.
>   /// @returns The inut set of points described in the frame referenced by
> ```

"inut" -> "input"

---

*[drake/multibody/rigid_body_tree.h, line 515 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqE2Hadx3XTNunVOnt:-KXqE2Hadx3XTNunVOnu:by1jzdh) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L515)):*
> ```objc
>   /// Given a set of `n` points provided as column entries in @p points in
>   /// `R^{3 x n}` this method computes a Jacobian matrix that relates each
>   /// point's velocity with the generalized coordinates time derivatives
> ```

"with" -> "with either"
"generalized coordinates' time derivatives" or, more comprehensibly, "time derivatives of generalized coordinates"

---

*[drake/multibody/rigid_body_tree.h, line 520 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEOBtFJQ4Bh-esh4t:-KXqEOBtFJQ4Bh-esh4u:b-vqpfu4) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L520)):*
> ```objc
>   ///
>   /// Defining `ndof = nq` if `in_terms_of_qdot = true` or `ndof = nv` if
>   /// `in_terms_of_qdot = false`, the Jacobian matrix lives in
> ```

"lives in" -> (perhaps) "is defined in"

---

*[drake/multibody/rigid_body_tree.h, line 523 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEc_fcfNxN3bULaB-:-KXqEc_fcfNxN3bULaB0:b-5nvoot) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L523)):*
> ```objc
>   /// `R^{3*n x ndof}` such that the linear velocity `v_i` of the `i-th`
>   /// point is computed as `v_i = [J * v].segment<3>(i)` or, in terms of
>   /// generalized coordinates time derivatives as
> ```

Same issue as above.

---

*[drake/multibody/rigid_body_tree.h, line 526 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEhsgRGFqBd2skcCr:-KXqEhsgRGFqBd2skcCs:bsm5h45) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L526)):*
> ```objc
>   /// `v_i = [J * qdot].segment<3>(i)`.
>   ///
>   /// Note: Many entries in this matrix may be zero for non-participating
> ```

I understand what you mean by "non-participating degrees of freedom" but I wouldn't expect the average user to understand.

---

*[drake/multibody/rigid_body_tree.h, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEshZ6g66Ih2dmhhO:-KXqEshZ6g66Ih2dmhhP:bj2f9ub) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L527)):*
> ```objc
>   ///
>   /// Note: Many entries in this matrix may be zero for non-participating
>   /// degrees of freedom. This might have an impact in the efficiency of your
> ```

"in the" -> "on the'

---

*[drake/multibody/rigid_body_tree.h, line 533 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqF04KMV1C5ur2vNF1:-KXqF04KMV1C5ur2vNF2:b-mmw0ei) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L533)):*
> ```objc
>   /// @param[in] points A matrix in R^{3 x n} containing n 3D column vectors
>   /// for each point to be transformed.
>   /// @param[in] from_body_or_frame_ind Frame identifier on which input points
> ```

Same issue as in previous function.

---

*[drake/multibody/rigid_body_tree.h, line 547 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqF92BByKt0mt0olho:-KXqF92BByKt0mt0olhp:bck8lgy) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L547)):*
> ```objc
>   /// @see transformPoints() to transform positions instead of velocities.
>   ///
>   /// The Jacobian in this method only allows to obtain the linear velocity
> ```

I'd say something like "The Jacobian returned consists only of the partial derivatives of the input points computed with respect to the degrees of freedom. It does not compute the partial derivatives of a frame orientation with respect to the degrees of freedom."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 1 of 2 files at r2.
Review status: 1 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.cc, line 1840 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr81xhw02zSHJjBuzl:-KXr81xiFazW4QRqZjFq:b-caazii) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1840)):*
> ```cpp
>                         in_terms_of_qdot, &v_or_q_indices);
> 
>   // Jomega, Jv live in R^{3 x ndof}.
> ```

Same issue as below about "lives in"

---

*[drake/multibody/rigid_body_tree.cc, line 1842 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8DLJKcvVFOiWoDzf:-KXr8DLJKcvVFOiWoDzg:b-xplqcd) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1842)):*
> ```cpp
>   // Jomega, Jv live in R^{3 x ndof}.
>   auto Jomega = J_geometric.template topRows<kSpaceDimension>();
>   auto Jv = J_geometric.template bottomRows<kSpaceDimension>();
> ```

BTW, consider something other than `Jv`, which looks at a glance to me like Jacobian * generalized velocity. Perhaps `Jxd`? Or `Jang` and `Jlin`?

---

*[drake/multibody/rigid_body_tree.cc, line 1844 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8SD9NhAyjpNN3ArS:-KXr8SD9NhAyjpNN3ArT:bg1y08h) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1844)):*
> ```cpp
>   auto Jv = J_geometric.template bottomRows<kSpaceDimension>();
> 
>   // J lives in R^{3*npoints x ndof}.
> ```

Same issue about "lives in"

---

*[drake/multibody/rigid_body_tree.cc, line 1870 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8qiFTW8b2WgaGnGo:-KXr8qiGkkX7kWIS_jtU:bvblvrq) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1870)):*
> ```cpp
>          it != v_or_q_indices.end(); ++it) {
>       // Since J_geometric relates generalized velocities (or generalized
>       // coordinates time derivatives) to Plucker vectors, the code below
> ```

Same genitive issue as below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Addressed reviews. PTAL

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.cc, line 1840 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr81xhw02zSHJjBuzl:-KXv99KDPx68EUr2x94x:bj1mne6) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1840)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same issue as below about "lives in"
</blockquote></details>

I think it is ok? is this a BTW comment?

---

*[drake/multibody/rigid_body_tree.cc, line 1842 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8DLJKcvVFOiWoDzf:-KXvA8IlupOODRi7ZHrq:b-8wu1is) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1842)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW, consider something other than `Jv`, which looks at a glance to me like Jacobian * generalized velocity. Perhaps `Jxd`? Or `Jang` and `Jlin`?
</blockquote></details>

BTW. I actually used `Jv`/`Jw` a lot in the past and I actually like it. I added some comments documenting what they are to clarify your point.

---

*[drake/multibody/rigid_body_tree.cc, line 1844 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8SD9NhAyjpNN3ArS:-KXvAL2_DAZAOYS_6SyZ:b-u47dhh) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1844)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same issue about "lives in"
</blockquote></details>

same answer.

---

*[drake/multibody/rigid_body_tree.cc, line 1870 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr8qiFTW8b2WgaGnGo:-KXvBDoJRELAvSKSjHum:b-a7wazo) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1870)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same genitive issue as below.
</blockquote></details>

sorry, "below" sometimes does not make much sense in Reviewable. Could you be more explicit? I did change what I think you meant?

---

*[drake/multibody/rigid_body_tree.h, line 465 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqD6udzlxn-3ZBVPCn:-KXv1Uc4fS1pDQg5HLIC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L465)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"in described in"
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_tree.h, line 470 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqDD5K3kZqYrBWRzdP:-KXv1lpSCqYBLYs5XHxm:b80qzgm) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L470)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Imagine that I'm looking up this function and I see that it takes an int for a body / frame identifier. What are valid values for that int? What are the meanings of the values? (0 refers to frame what? 1 refers to frame what?). Answering this might also answer my question: why does this method exist here? Put another way: points can be transformed between frames without needing to know anything about rigid bodies. So why is this function in RigidBodyTree? 
</blockquote></details>

Yes, I hate this API but in this PR I am just limiting myself to document what I learned it does.
The transformations depends on state (cached here) because these frames id's always reference a frame that is attached to a RigidBody, which transformation to world depends on its state (cached here).
I updated the doc to reflect this.
Unfortunatelly there are not good examples of RigidBodyFrame usage in the C++ API. I added TODO.
Also `RigidBodyFrame` is used when create `LoopJoins` but I don't want to reference that code because it will be upgrade when I implement a general `Constraint` class.

---

*[drake/multibody/rigid_body_tree.h, line 474 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqD2egJCIsi4He9ovu:-KXv648ChOksonRu1XAJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L474)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"inut" -> "input"
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_tree.h, line 515 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqE2Hadx3XTNunVOnt:-KXv6xD25hcmxwKq--oY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L515)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"with" -> "with either"
"generalized coordinates' time derivatives" or, more comprehensibly, "time derivatives of generalized coordinates"
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_tree.h, line 520 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEOBtFJQ4Bh-esh4t:-KXv7MIE8ggsQH-wfTVw:b-t2kt7x) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L520)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"lives in" -> (perhaps) "is defined in"
</blockquote></details>

"lives in" is common math terminology right? It reads like "the matrix lives in this/that space"

---

*[drake/multibody/rigid_body_tree.h, line 523 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEc_fcfNxN3bULaB-:-KXv7_JflUWxww2KywWA:b-urb88d) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L523)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same issue as above.
</blockquote></details>

same answer

---

*[drake/multibody/rigid_body_tree.h, line 526 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEhsgRGFqBd2skcCr:-KXv7ipv5UVvwmzLbJND:blo944w) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L526)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I understand what you mean by "non-participating degrees of freedom" but I wouldn't expect the average user to understand.
</blockquote></details>

I thin the "average" user calling a method like `transformPointsJacobian` should understand this. But if you don't agree please tell me, how would you explain that? or would make this not a developer comment (not a Doxygen doc)?

---

*[drake/multibody/rigid_body_tree.h, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEshZ6g66Ih2dmhhO:-KXv85HMLVYHqw3jDNTd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L527)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"in the" -> "on the'
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_tree.h, line 533 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqF04KMV1C5ur2vNF1:-KXv8IkOk1gpazI0FDKM:b-htvxtr) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L533)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same issue as in previous function.
</blockquote></details>

I added the same explanatory text here also. PTAL.

---

*[drake/multibody/rigid_body_tree.h, line 547 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqF92BByKt0mt0olho:-KXv90lecNd0YObiKp5f:b-bhzh0e) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L547)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I'd say something like "The Jacobian returned consists only of the partial derivatives of the input points computed with respect to the degrees of freedom. It does not compute the partial derivatives of a frame orientation with respect to the degrees of freedom."
</blockquote></details>

I like it. I added it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvBjmlGAAmu77kWpE2:bvvuhwq)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@jwnimmer-tri for platform please

---

Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvBtIjvsXUL88YTi0v:b-1t4l8z)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
So this is a bunch of new comments that I won't be able to understand.  I can rubber-stamp it LGTM if you like, but would you rather have Sherm be your platform reviewer?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvMmN8R9PWW6B8tjX7:btrlepi)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Well <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> in any case if you want to merge.  You can assign @sherm1 additionally if you prefer.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvWxuKy8yzDnlUaV0w:b-oossst)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I was just following protocol and I thought having @edrumwri taking a look was enough but, actually I'd like @sherm1 to take a look at the introduction of some new monogram notation please.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvXIZDXCWSq_diycOW:b-qk24py)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
It's fine to have me platform-review it (and I did), but to be clear, the "protocol" doesn't mean that you _must_ assign to me to the review.  The protocol is that all else being equal, I will be on call to review things quickly today, but if your code needs a platform reviewer other than me, you are still invited to assign it to them at your discretion.

amcastro-tri commented:
Sounds good, thank you @jwnimmer-tri. +@sherm1 for platform (/feature) review please. @sherm1 in particular see if I got your monogram notation right. 

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXvYxhRRuVTVRNBDF7l:b-9tco9h)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.cc, line 1840 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXr81xhw02zSHJjBuzl:-KXvYH73g12PcjISJeAC:bw0656m) ([raw file](https://github.com/robotlocomotion/drake/blob/ca31e6165a41c472507ed71950046f6625b4b560/drake/multibody/rigid_body_tree.cc#L1840)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think it is ok? is this a BTW comment?
</blockquote></details>

Office consensus says "element of" or simply "in" instead of "lives in".

---

*[drake/multibody/rigid_body_tree.cc, line 1876 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvyHwLpHvRAmJOsWWn:-KXvyHwLpHvRAmJOsWWo:bxey6dc) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.cc#L1876)):*
> ```cpp
>       // derivatives of generalized coordinates) to Plucker vectors, the code
>       // below makes the transformation from the origin's linear velocity
>       // (actually the base linear velocity) to the particular point linear
> ```

Why is this elaboration necessary? Now I feel like I'm missing something.

---

*[drake/multibody/rigid_body_tree.h, line 470 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqDD5K3kZqYrBWRzdP:-KXvoz-n_U9W1AB6QlxK:bc63nha) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L470)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Yes, I hate this API but in this PR I am just limiting myself to document what I learned it does.
The transformations depends on state (cached here) because these frames id's always reference a frame that is attached to a RigidBody, which transformation to world depends on its state (cached here).
I updated the doc to reflect this.
Unfortunatelly there are not good examples of RigidBodyFrame usage in the C++ API. I added TODO.
Also `RigidBodyFrame` is used when create `LoopJoins` but I don't want to reference that code because it will be upgrade when I implement a general `Constraint` class.
</blockquote></details>

I'd say that the documentation can't be complete without answering my first two questions...

---

*[drake/multibody/rigid_body_tree.h, line 526 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEhsgRGFqBd2skcCr:-KXvqQotlI767ki6JldV:b-ssnisc) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L526)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I thin the "average" user calling a method like `transformPointsJacobian` should understand this. But if you don't agree please tell me, how would you explain that? or would make this not a developer comment (not a Doxygen doc)?
</blockquote></details>

Just flesh it out a bit or perhaps give an example. You could say that the degree of freedom does not alter the given link.

---

*[drake/multibody/rigid_body_tree.h, line 546 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvqIuJkecFt6umswFG:-KXvqIuJkecFt6umswFH:b-14ovn8) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L546)):*
> ```objc
>   ///   `vi`: The linear velocity of the i-th point in `points`.
>   ///
>   /// The Jacobian matrix lives in `R^{3*n x ndof}` and relates to the linear
> ```

Sorry Alejandro, but "relates to" is imprecise. Look at the wording I gave you for the other Jacobian description.

---

*[drake/multibody/rigid_body_tree.h, line 564 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvqtVrCx-ZB6fHRkKH:-KXvqtVrCx-ZB6fHRkKI:bj60b9k) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L564)):*
> ```objc
>   /// computed with this Jacobian are described i.e. `v_i`'s are described in
>   /// to_body_or_frame_ind.
>   /// @param[in] in_terms_of_qdot if `true` the returned Jacobian allows to
> ```

"allows to compute" -> "allows computing"

---

*[drake/multibody/rigid_body_tree.h, line 571 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvqzLrrKZIeWvCvVro:-KXvqzLsdJ0IIzsMWlU6:b-4h1vi4) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L571)):*
> ```objc
>   /// `in_terms_of_qdot = true` or `ndof = nv` if `in_terms_of_qdot = false`.
>   ///
>   /// @see transformPoints() to transform positions instead of velocities.
> ```

positions of what? velocities of what?

---

*[drake/multibody/rigid_body_tree.h, line 577 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvxjFbf7qYHkmsUD2y:-KXvxjFbf7qYHkmsUD2z:b-seijph) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L577)):*
> ```objc
>   /// compute the partial derivatives of a frame orientation with respect to
>   /// the degrees of freedom. @see geometricJacobian() to obtain the Jacobian
>   /// relating generalized velocities (or generalized coordinates time
> ```

genitive issue

---

*[drake/multibody/rigid_body_tree.h, line 627 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvw701AW54goqiicbL:-KXvw701AW54goqiicbM:b-seijph) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L627)):*
> ```objc
>   /// respect to frame `O` exressed in frame `E`.
>   /// The Jacobian can be computed in terms of generalized
>   /// coordinates time derivatives (i.e. `V_OB_E = Jg * qdot`) or in terms of
> ```

genitive issue

---

*[drake/multibody/rigid_body_tree.h, line 637 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvsGmX1QkCuMkDId6T:-KXvsGmYmlTeYqxIUUCA:b-8uulmg) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L637)):*
> ```objc
>   ///   `E`: the frame referenced by expressed_in_body_or_frame_ind.
>   ///
>   /// The Jacobian matrix `Jg` lives in `R^{6 x ndof}` and relates to the
> ```

More "lives in"

---

*[drake/multibody/rigid_body_tree.h, line 640 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvwBQjWYQ8ANE-srLn:-KXvwBQjWYQ8ANE-srLo:b-seijph) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L640)):*
> ```objc
>   /// Plucker vector represenation of the spatial velocity `V_OB` by:
>   ///   `V_OB_E = Jg * v` (in terms of generalized velocites)
>   ///   `V_OB_E = Jg * qdot` (in terms of generalized coordinates
> ```

genitive issue

---

*[drake/multibody/rigid_body_tree.h, line 644 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvw9jy9BioH4DWTU5w:-KXvw9jy9BioH4DWTU5x:b-cnalfk) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L644)):*
> ```objc
>   ///
>   /// Note: Many entries in this matrix may be zero for non-participating
>   /// degrees of freedom. This might have an impact in the efficiency of your
> ```

"in" again

---

*[drake/multibody/rigid_body_tree.h, line 657 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvwFzryL8kGucTtkEN:-KXvwFzryL8kGucTtkEO:b-tear4z) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L657)):*
> ```objc
>   /// @returns The Jacobian matrix `Jg` in `R^{6 x ndof}`.
>   ///
>   /// @see transformPoints() to transform positions instead of velocities.
> ```

"positions instead of velocities" -> "transform points (instead of velocities) between frames" (or some other replacement that is more precise)

---

*[drake/multibody/rigid_body_tree.h, line 658 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvwZ8WKzG7aAjS-biU:-KXvwZ8WKzG7aAjS-biV:bmwz9aw) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L658)):*
> ```objc
>   ///
>   /// @see transformPoints() to transform positions instead of velocities.
>   /// @see transformPointsJacobian() to obtain a Jacobian matrix relating
> ```

"relating" is insufficiently precise. You either want to provide the definition of the Jacobian matrix as a group of partial derivatives of a function with multiple outputs or you can say that the Jacobian transforms one quantity to another.

---

*[drake/multibody/rigid_body_tree.h, line 667 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvwtJMAKzYuR0OtlFT:-KXvwtJMAKzYuR0OtlFU:b-6lywe7) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L667)):*
> ```objc
>   /// Frame" which is the equivalent of Plucker vectors in Section 2.5, p.31.
>   /// Velocity recursion using these frames is discussed as part of
>   /// Exercise 3.4, p. 47.
> ```

Would you be willing to provide the reference for Jain's book in the following format (to be consistent with my code)?
[Jain 2010]  A. Jain. Robot and Multibody Dynamics. Springer. New York, 2010.

---

*[drake/multibody/rigid_body_tree.h, line 1136 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXvro1uNH6Ph8LikACX:-KXvro1uNH6Ph8LikACY:b-gy2meq) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L1136)):*
> ```objc
>   /// Returns the index of the RigidBody associated with the frame
>   /// represented by body_or_frame_id. Therefore, this method always returns
>   /// the index of a RigidBody.
> ```

Where does this index come from, Alejandro?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Partial review, @amcastro-tri -- PTAL. This is really hard to review because there are lots of changes I'd like to see. I tried to give some examples for two of the functions. I think the way to proceed would be to discuss the documentation for just one function until we're all happy with it, then follow the same style for the next function, review, etc.

The documentation as written (for the methods I looked at) isn't complete or precise and doesn't make enough use of mathematics to make the operations unambiguous. Only the mathematical description can be fully precise for some of these methods. We can try to help out non-expert callers also, but the really important part now that *you* understand these functions is to get that understanding written down unambiguously. If we expect non-expert users for these methods (is that even conceivable?), then we can attempt to translate the math to English but the critical part is the math because the English is going to be imprecise.

---

Review status: all files reviewed at latest revision, 29 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.h, line 517 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwySsoCedgjt0dWiFJ:-KXwySsoCedgjt0dWiFK:burcev1) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L517)):*
I don't think this completely describes the function. 
- it returns a Jacobian that maps generalized velocities to point velocities
- it is given a frame F, and a set of point locations measured from and expressed in F
- the points are assumed to be fixed in frame F [is that true?]
- it is given a frame T with respect to which the resulting point velocities are measured [is that true?] Frame T is also used to express the resulting velocity vectors [is that true?] Maybe the velocities are returned in the World frame but expressed in T?
- the returned Jacobian should then be called J_TF to incorporate the frames
- the function accepts points that have a different scalar type than the resulting Jacobian's scalar type
- it allows generalized speeds to be given in terms of v or qdot

I would like to see something with this level of precision (I'm guessing here what the function actually does -- please fix!):
> Let `x_F` be the vector of locations of points (1) fixed in frame F, (2) measured from origin `Fo` of the F frame, and (3) expressed in the F frame. The goal is to produce the time derivative of `x_F`, taken in some other frame T. Let's call that derivative `xd_TF`. Then define` J_TF` such that `xd_TF=J_TF*v` and define`Jq_TF` such that` xd_TF=J_TF*qdot`. 

Given something like that, you could then say unambiguously that the function returns either J_TF or Jq_TF depending on the `in_terms_of_qdot` flag.

We also have the identities `Jq_TF=J_TF*N+` and `J_TF=Jq_TF*N` where N is the kinematic differential equation mapping such that `qdot=N*v` and `N+` its left pseudoinverse. That might be worth pointing out here since I think it helps to clarify the two options.

---

*[drake/multibody/rigid_body_tree.h, line 520 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXqEOBtFJQ4Bh-esh4t:-KXwtoOvx-1J17H-lUiI:b-gvx0ck) ([raw file](https://github.com/robotlocomotion/drake/blob/63ae0effbda2dd080af1dfaa830acac1ba82d72c/drake/multibody/rigid_body_tree.h#L520)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

"lives in" is common math terminology right? It reads like "the matrix lives in this/that space"
</blockquote></details>

Math terminology would be M∈R^mxn. You could use the "element-of" character in the comment, or you could say "M in R^mxn" but "lives in" seems odd. 

---

*[drake/multibody/rigid_body_tree.h, line 466 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwpiRSkGVNb2YlKVU-:-KXwpiRTGWxExNgcEq8O:b4m3pvb) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L466)):*
> ```objc
>   /// Given a set of points in the frame referenced by
>   /// @p from_body_or_frame_ind this methods returns the same set of points
>   /// described in the frame referenced by @p to_body_or_frame_ind.
> ```

To be more precise, how about: "Given a set of point locations measured and expressed in "from" frame F, this method returns the same locations in space but measured and expressed in "to" frame T. The frame indexes are supplied as parameters."

---

*[drake/multibody/rigid_body_tree.h, line 474 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwqyp0D37KaVKrLeyy:-KXwqyp0D37KaVKrLeyz:b4h5tk6) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L474)):*
> ```objc
>   /// the result from `RigidBodyFrame::get_frame_index()`. Therefore the
>   /// result is state dependent, which gets reflected in the fact that his
>   /// method takes a KinematicsCache `cache` objects as an input.
> ```

BTW, must `doKinematics()` have already been called on KinematicsCache so that it is up to date?

---

*[drake/multibody/rigid_body_tree.h, line 478 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwr8ssbOBSCiQqCeiz:-KXwr8ssbOBSCiQqCej-:b-icegwl) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L478)):*
> ```objc
>   /// @param[in] cache KinematicsCache object.
>   /// @param[in] points A matrix in R^{3 x n} containing n 3D column vectors
>   /// for each point to be transformed.
> ```

BTW, call this parameter `points_F`? "A matrix ... to be transformed, measured and expressed in frame F."

---

*[drake/multibody/rigid_body_tree.h, line 479 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwrRvY9NJv_4wTTNYV:-KXwrRvY9NJv_4wTTNYW:b-qdrg2n) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L479)):*
> ```objc
>   /// @param[in] points A matrix in R^{3 x n} containing n 3D column vectors
>   /// for each point to be transformed.
>   /// @param[in] from_body_or_frame_ind Frame identifier on which input points
> ```

"Frame identifier for the "from" frame F in which the input points are given."

---

*[drake/multibody/rigid_body_tree.h, line 482 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwrYyGmQ0yJwSne0uM:-KXwrYyGmQ0yJwSne0uN:b-wfow61) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L482)):*
> ```objc
>   /// are described.
>   /// @param[in] to_body_or_frame_ind Frame identifier on which output points
>   /// are described.
> ```

"Frame identifier for the "to" frame T in which the output points are given."

---

*[drake/multibody/rigid_body_tree.h, line 483 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwrhTfsqoRce1RmAAr:-KXwrhTgtDzNM_koYLi0:b47w35q) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L483)):*
> ```objc
>   /// @param[in] to_body_or_frame_ind Frame identifier on which output points
>   /// are described.
>   /// @returns The input set of points described in the frame referenced by
> ```

"returns The same locations as represented by the input points, but measured and expressed in frame T, and with the data types converted to `Scalar`.

---

*[drake/multibody/rigid_body_tree.h, line 487 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwp0wO6Nd8xwIoXtXm:-KXwp0wO6Nd8xwIoXtXn:berxus9) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L487)):*
> ```objc
>   // TODO(amcastro-tri): write similar C++ test to the one in
>   // drake/bindings/python/pydrake/test/testPR2IK.py showing the use of
>   // RigidBodyFrame.
> ```

BTW, should have `@tparam Scalar` and `@tparam DerivedPoints` to document the two template arguments.
Also, should say something about the type conversion that goes on here from whatever kind of matrix holds the input points to an output 3xn matrix of scalars.

---

*[drake/multibody/rigid_body_tree.h, line 504 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KXwsaI1_0VI4uj9W6rN:-KXwsaI1_0VI4uj9W6rO:b-mkhqn2) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L504)):*
> ```objc
>     auto relative_transform =
>         relativeTransform(cache, to_body_or_frame_ind, from_body_or_frame_ind);
>     return relative_transform * points.template cast<Scalar>();
> ```

With the comments modified as above, you could replace this with R_TF and write `R_TF*points_F` to show that the transformation from F to T is correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXx5jMnp2xFuMNOKMeQ:b3t1xp3)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@amcastro-tri, I noticed you distinguish Plucker vector velocities from spatial velocities. I thought those were identical and that only the acceleration vectors differ by a cross product. Are you referring only to the choice of coordinate frame? Spatial velocity vectors don't have to be measured or expressed in the World frame.

---

Review status: all files reviewed at latest revision, 29 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KXx9UCrsYzkhLolNEhf:ba4rxzp)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you for taking a look @sherm1. Please take a look at my comment below referencing a one of Jain's book chapters. 
Also, to agree on a style to document these methods, I propose we start with `geometricJacobian` which seems to introduce all the elements needed in the other methods.
My mistake here was trying to write a documentation for general users in plain English when in reality only advanced users knowing what they are doing will call these (in contrast to simpler-level user methods like the ones I propose in #4328). I agree I should therefore describe them in precise mathematical terms. 

---

Review status: all files reviewed at latest revision, 30 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.h, line 666 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KY-HIP3UrNtOFuLtonW:-KY-HIP5DixeuTSC83BT:b-egjd7c) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L666)):*
> ```objc
>   /// A. Jain mentions the notion of "Inertially Fixed Velocity Reference
>   /// Frame" which is the equivalent of Plucker vectors in Section 2.5, p.31.
>   /// Velocity recursion using these frames is discussed as part of
> ```

@sherm1, could you look at this reference? In my understanding what Jain was referring there as spatial vectors described in a "Inertially Fixed Velocity Reference Frame" was equivalent to Plucker vectors. But I might be wrong on that one. 
However, something I am 100% sure, is that the math followed by these methods is the one described in that Chapter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-:-KY-HnvGRgyZITsl439Q:b-lf17n9)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/multibody/rigid_body_tree.h, line 666 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KY-HIP3UrNtOFuLtonW:-KY0QCHoqBHUBTnMEjgj:b-u2fji4) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L666)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

@sherm1, could you look at this reference? In my understanding what Jain was referring there as spatial vectors described in a "Inertially Fixed Velocity Reference Frame" was equivalent to Plucker vectors. But I might be wrong on that one. 
However, something I am 100% sure, is that the math followed by these methods is the one described in that Chapter.
</blockquote></details>

An aside: Jain's comment at the bottom of page 8 is interesting: "The spatial notation used here is based on references [includes Featherstone 2008]. A slight variant, based on the classical concepts of screws and wrenches, was used originally in [cites Featherstone's older stuff]."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 30 unresolved discussions.

---

*[drake/multibody/rigid_body_tree.h, line 666 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4311#-KY-HIP3UrNtOFuLtonW:-KY0T2WZQqtJMz048mUN:b-4z42hj) ([raw file](https://github.com/robotlocomotion/drake/blob/4c34e223d11c86cd00f447df1ad8dfbb3732199f/drake/multibody/rigid_body_tree.h#L666)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

An aside: Jain's comment at the bottom of page 8 is interesting: "The spatial notation used here is based on references [includes Featherstone 2008]. A slight variant, based on the classical concepts of screws and wrenches, was used originally in [cites Featherstone's older stuff]."
</blockquote></details>

Thanks, Alejandro -- I hadn't noticed that correspondence before!
Featherstone 2008's eq 2.4 on page 12 shows that the Plucker velocity vector is the same as Jain's spatial velocity, and 2.9 shows the same thing for spatial forces. Eq 2.48 on page 29 shows that the Plucker acceleration is *not* the same as Jain's spatial acceleration due to the definition of the linear velocity term. I *think* that is like the "inertially fixed velocity reference" you mentioned -- the velocity is of a "flow" through a point rather than of the point itself. (That is, the reference point is moving w.r.t. the body frame, adding an extra term to the time derivative.) So I agree with your analysis but I *really, really* hate that the Plucker linear acceleration term is not the acceleration of a point!

If possible, where you have to talk about R^6 Plucker vectors to describe the existing code, can you also mention where it is different from ordinary R^3 x R^3 spatial vectors (like: to get the actual acceleration subtract `w x v` from this vector, or whatever)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4311)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-xenial-clang-experimental please

jwnimmer-tri commented:
@amcastro-tri Is this PR still in progress?  I am wondering if its been overtaken by events, and could be closed?

amcastro-tri commented:
overtaken for events and not a high priority. However I'd like to merge this so we don't loose the documentation I wrote. I'll cleanup by the end of this week to have another round of review. Thanks @jwnimmer-tri

jwnimmer-tri commented:
@amcastro-tri re-ping.

jamiesnape commented:
@drake-jenkins-bot retest this please

jamiesnape commented:
@drake-jenkins-bot linux-xenial-gcc-experimental-open-source please

jwnimmer-tri commented:
@amcastro-tri re-ping.

amcastro-tri commented:
I will make sure this is considered per #5013.
